### PR TITLE
Anynomous functions now allows exception handling with the do... end syntax

### LINKED ISSUE #86: Allow exception handlers on anonymous function clauses.
_This issue has been marked as a "Good First Issue"! If you'd like to tackle this issue as your first contribution to the project, be sure to read the Get Involved section of the README for some help with how to get started._

For consistency with normal method definitions and blocks, anonymous function clauses using the `do...end` syntax should be allowed to define exception handlers:

```myst
func = fn
  ->(2) do
    raise "woops"
  rescue
    :rescued
  end
end

func(2) #=> :rescued
```

The semantics of this should already be in place in the interpreter. The only necessary change should be in the parser, and will likely just be an addition to [`parse_anonymous_function`](https://github.com/myst-lang/myst/blob/master/src/myst/syntax/parser.cr#L299) around this area to call `parse_exception_handler`:

https://github.com/myst-lang/myst/blob/3fbbdc55052318071c8661d9b6d4898e337169c0/src/myst/syntax/parser.cr#L341-L344

For inspiration, you can look at how `parse_optional_block` implements the same thing:

https://github.com/myst-lang/myst/blob/3fbbdc55052318071c8661d9b6d4898e337169c0/src/myst/syntax/parser.cr#L849-L859

There's not too much more to this change. Be sure to add some parser specs around this new syntax. The [exception handler specs for blocks](https://github.com/myst-lang/myst/blob/master/spec/syntax/parser_spec.cr#L2024) should be fairly easy to adapt to use anonymous functions instead.

If you have any questions, feel free to ask in the #help channel in [the discord server](https://discord.myst-lang.org/) or message me directly. Good luck! :)

### PR DESCRIPTION
**Pull request by Jens0512**
This was just appending two lines to an already existing method, so i hope i didn't get anything wrong.

~I took a look at the `parser_spec.cr` file, but i have to admit it scared me away.~ EDIT: Specs added

I believe this patch resolves #86

### CODE REVIEW

**Comment by Jens0512** [comment]
With this something like 
```ruby
func = fn
  ->(2) do
    raise "woops"
  rescue
    :rescued
  end
end
``` 
will work as one would expect (`func(2) # => :rescued`)

As you might notice, my editor (sublime) took the liberty of running crystal format on `parser.cr` couple of times.

**Comment by Jens0512** [comment]
Specs coming up

**Comment by Jens0512** [comment]
Seems something is wrong with travis again, not the first time i've seen this happen.

File: spec/syntax/parser_spec.cr [hunk 1]
```diff
@@ -2474,8 +2474,29 @@ describe "Parser" do
...
+      }
+    end
+  )
+
+  # Expection handling in anonymous functions with do... end syntax is allowed
+  it_parses %q(
+    fn
+      ->() do
+        raise :error
+        rescue
```
Comment (faultyserver) [inline spec/syntax/parser_spec.cr]: `rescue` and `end` here should be indented to match the `->()` that starts the clause (similar to how they match with `def` or other blocks).

File: spec/syntax/parser_spec.cr [hunk 2]
```diff
@@ -2474,8 +2474,29 @@ describe "Parser" do
...
+          :rescue
+        end
+    end
+  ),                        AnonymousFunction.new([Block.new(body: ExceptionHandler.new(Raise.new(l(:error)), [Rescue.new(l(:rescue))]))])
```
Comment (faultyserver) [inline spec/syntax/parser_spec.cr]: It'd be nice to cover more cases here as well: rescue with parameters, multiple rescues, rescue with ensure, etc. These may fit better with the rest of the [exception handling specs](https://github.com/myst-lang/myst/blob/master/spec/syntax/parser_spec.cr#L1687-L2056). In fact, just duplicating all of the specs with `def` and replacing it with `fn ->() do` would probably do the trick.

File: src/myst/syntax/parser.cr [hunk 3]
```diff
@@ -679,10 +680,10 @@ module Myst
...
         return parse_postfix(call)
-      # While `parse_var_or_call` can distinguish Calls for string identifiers,
-      # its isolated scope (just the current token) means it cannot be
-      # responsible for parsing Calls from arbitrary expressions. However,
-      # since these are just postfix expressions, those can be handled here.
+        # While `parse_var_or_call` can distinguish Calls for string identifiers,
+        # its isolated scope (just the current token) means it cannot be
+        # responsible for parsing Calls from arbitrary expressions. However,
+        # since these are just postfix expressions, those can be handled here.
       when start = accept(Token::Type::LPAREN)
```
Comment (faultyserver) [inline src/myst/syntax/parser.cr]: I'm generally not too opposed to the Crystal formatter, but this comment block relates to the `when` on line 687, so its indentation should stay in line with it.

**Review by faultyserver — COMMENTED** [review]
Nice work! I'd like to see some better spec coverage before merging, though, and the `crystal format` changes are a bit much for my liking.

I think you mentioned you used Sublime? You can disable the auto-format-on-save with a configuration option to make things a bit easier (it will still be accessible through the command palette).

File: src/myst/syntax/parser.cr (hunk 3 above)
Comment (Jens0512) [reply src/myst/syntax/parser.cr]: Sorry about all the auto-formatting. I very much agree that it should be off, found out i could turn it off (which i did) shortly after committing the auto-formatted changes.

File: spec/syntax/parser_spec.cr (hunk 2 above)
Comment (Jens0512) [reply spec/syntax/parser_spec.cr]: Its quite a lot, and is it not `parse_exception_handler` which is being used in the exception handling specs? 

That is just the same which is used in the anonymous functions, I mean, I have no experience in this, so i just wonder why this is different?

Comment (faultyserver) [reply spec/syntax/parser_spec.cr]: The goal is to have complete coverage of the syntax, without any assumptions about how the parser is implemented. So, while the parser is currently using `parse_exception_handler` in both cases, that isn't _necessarily_ guaranteed. It's also possible that a future change to `parse_exception_handler` could add some stateful condition that causes it to work properly on exception handlers for blocks, but not anonymous functions, etc. Without tests for all the possibilities on anonymous functions, we wouldn't be able to catch that immediately.

It's certainly tedious to start with, but the complete test coverage means we can much more confidently make larger changes to the parser without worry of missing or breaking something.

Comment (Jens0512) [reply spec/syntax/parser_spec.cr]: Gotcha! I'll see to it

**Comment by Jens0512** [comment]
**NB:** Not done yet

**Comment by faultyserver** [comment]
This looks pretty solid to me. Do you still have something else you want to add?

**Comment by Jens0512** [comment]
@faultyserver No, i think this is pretty much good to go

**Pull request merged**